Add sitemap

cb372 opened the pull request:
Adds new routes, models and actor for exposing a sitemap for the documentation pages. This allows us to set lower priorities for documentation relating to older versions of Play, allowing some control over where they appear in search engine results. See issue #8 for more details.

I'm open to suggestions on how to calculate the priorities for various versions. The policy I implemented is pretty arbitrary and kludgy. See the code comments for details on the logic, and take a look at the test to see how it ranks various versions. (We have access to the release dates for each version in `playReleases.json`, so we could maybe prioritise recently released versions, even if they are not numerically the newest.)

A sitemap entry looks like this:

```
<url>
  <loc>https://www.playframework.com/documentation/2.4.x/JavaTestingWebServiceClients</loc>
  <priority>0.73</priority>
  <xhtml:link rel="alternate" hreflang="tr" href="https://www.playframework.com/documentation/tr/2.4.x/JavaTestingWebServiceClients"/>
  <xhtml:link rel="alternate" hreflang="fr" href="https://www.playframework.com/documentation/fr/2.4.x/JavaTestingWebServiceClients"/>
  <xhtml:link rel="alternate" hreflang="bg" href="https://www.playframework.com/documentation/bg/2.4.x/JavaTestingWebServiceClients"/>
</url>
```

(For some reason on my local machine the app thinks the 2.4.x docs are translated into French and Bulgarian as well as Turkish, but the live site disagrees.)

This PR adds 2 new routes:

* `/sitemap.xml` - the sitemap itself
* `/robots.txt` - a robots file to point crawlers to the sitemap

richdougherty commented:
Hi @cb372. Firstly, thanks a lot for doing this, it's very helpful.

A few thoughts:

1. What shall we do about pages that aren't listed in the sitemap? Some pages that aren't listed, like the home page and download page, are really important, so we want them to have a priority at least as high as the docs. Some [docs for minor versions](https://www.playframework.com/documentation/2.3.1/GzipEncoding) aren't listed in the sitemap either, but they're not that important. They should have a priority lower than the docs listed in the sitemap.
2. In terms of priorities I think we should have:
  * `2.4.x > 3.0.x` because 3.0.x doesn't have a stable version yet
  * `2.4.x > 2.3.x` because newer versions are more important
  * `2.4.x > 2.4.2` because branches (2.4.x) are more important than specific versions (2.4.1)

    This is basically the same as what you had, except for the `2.4.x > 3.0.x` rule. You can determine that `2.4.x` is the main version by looking in the `Documentation` object under the `default.defaultVersion` property.
3. Did you think about splitting up the sitemap into different files and using a [sitemap index](http://www.sitemaps.org/protocol.html#index) to tie them together?

Here's one way of handling page priorities for pages that aren't in the documentation sitemap:

* Include the home page, download page, etc in a separate sitemap and give them a priority of 1.0. This sitemap could just be some static content.
* Include the default version of the docs in a documentation sitemap with a priority of 0.9. 

  It's not necessary to include any other versions of the docs in the sitemap, but it wouldn't hurt to do so, since you've already written the code! If you include other versions then their priorities should be above 0.5 and below 0.9.
* Let the rest of the site, including all other docs, have a priority of 0.5. This is the default priority so there's no need to list them in a sitemap.

cb372 commented:
Thanks @richdougherty, your suggestions sound sensible to me. I'll hardcode the important non-documentation pages into a separate sitemap and set up a sitemap index. Should have time to fix this over the weekend.

richdougherty commented:
Thanks @cb372!

cb372 commented:
Added a static sitemap for the non-documentation pages. Also simplified the documentation sitemap: it now includes only the docs for the latest version (2.4.x), with priority 0.9.

jroper commented on app/actors/SitemapGeneratingActor.scala:
```diff
@@ -0,0 +1,21 @@
+  import SitemapGeneratingActor._
+
+  def receive = {
+    case GenerateSitemap(documentation) =>
```
The sitemap could be cached based on an xor of the hashes of the translations - if that changes, regenerate the sitemap, otherwise, do nothing.

jroper commented:
@cb372 Thanks! This is awesome!

richdougherty commented:
This looks good to me now. @cb372, thanks for making the changes.

@jroper, I see you have one comment about caching. Do you want to merge this when you're happy?

richdougherty commented:
Looks like @jroper has deployed this now: https://www.playframework.com/robots.txt

jroper commented:
Yep, but Google has a problem with it, it's saying the `sitemap-main.xml` is not valid. That appears to be because it's using a top level element of `sitemap` rather than `urlset`.  I'll change it and see if that fixes it.

jroper commented:
Ok, I've updated, now to wait for Google to try and index again and see if the problem is solved.

richdougherty commented:
If this doesn't work, then the next thing we can try is adding links from each page of the old docs pointing to the latest docs. This should help both users and Google find the right page. Fingers crossed that the sitemap works though…

jroper commented:
I think it may be worth putting a small, noticeable, but not obnoxious warning at the top of each page saying that you're not viewing the current version of docs (and be specific, eg you're viewing the docs for a not yet released version, or for an old version) with a link to the current version of that page if it exists, otherwise no link.

richdougherty commented:
Yes, exactly. Something like how Django does it, except maybe have the warning under the page heading and maybe link to the equivalent page in the latest version of the docs rather than just having a warning.

* https://docs.djangoproject.com/en/dev/
* https://docs.djangoproject.com/en/1.6/
